### DISCUSSION TITLE
Numpad的bug

### DISCUSSION THREAD

**Discussion by 1111HGH** [Jagailo/YetAnotherKeyDisplayer#22] [Q&A] [answered]
Hello, I have used this 'YAKD' for a few months. It's very good, but I find a bug. I am from China. I can use accurate Chinese to show the bug accurately and completely, but my English may not as accurate as that.
就是有一个问题啊，关于键盘的”小键盘“那块地方，就是键盘最右边那些数字键，关于”打开小键盘“按键是有问题的，我举一个例子吧，是这样的，我同时按下"Num Lock"和”Numpad 6“然后松开”Num Lock“，再松开"Numpad 6"，这个时候你的YAKD面板上就会持续显示”Numpad 6“，直到再点一次”NumLock“和"Numpad 6"这样才能修复，而且这个bug是不管在”Num Lock“是否开启的时候都会出现的bug，另外一种情况就是显示，"Right"这个时候就是依次点击”Num Lock“和”Numpad 6“或者点击”上下左右“的”右→“才能恢复，希望能修改。

不过不得不说，你做的软件是真的好用，我在中国搜索引擎上搜过很多很多，效果都没你这个好！（对于一个弹键盘的人来说https://www.bilibili.com/video/BV1Zq4y1U7YD

There is a problem, The "keypad" part of the keyboard is the number keys on the right side of the keyboard. There is a problem with the "Open Keypad" key. Let me take an example. Here, I press "Num Lock" and "Numpad 6" at the same time, release "Numpad Lock", and then release "Numpad 6". At this time, "Numpad 6" will continue to be displayed on your YAKD panel until you click "Numpad 6" and "Numpad 6" again In this way, the bug can be fixed, and it is a bug that will appear no matter whether "Num Lock" is turned on or not. The other way is to show that "Right" at this time is to click "Num Lock" and "Numpad 6" in turn, or click "Right →" of "Up, Down, Left, Right" to recover. I hope it can be modified.

However, I have to say that your software is really useful. I have searched many Chinese search engines, and the results are not as good as yours! (For a person who plays the keyboard https://www.bilibili.com/video/BV1Zq4y1U7YD

**Comment by Jagailo** [comment, marked as answer]
Thank you for your kind words, I'm glad you're happy with my modest program. I have already encountered a similar problem, but, unfortunately, I don't know what to do with it.

The fact is that I get information about the keyboard keys from the Windows API. This information is presented in the form of events. For example, _key A is down_ -> _key A is up_, and so on. That is, I don't have information for all keys currently pressed. For example, you pressed Numpad 6, the program received a key press event and displayed it in the window. Then you pressed Numlock and released Numpad 6, but the number block was already disabled. The program didn't receive the event Numpad 6 is up and doesn't know that the key should be removed from the window.

This is a limitation of the physical keyboard and Windows. This is often used in games when you need to hold down a key for a very long time, then bind it to Numpad X, press the key and press Numlock. The key will always be pressed.

I could use another way to track the keys, but that requires the app to be in focus. And this isn't suitable for displaying keys in games.

So I can't do anything about it.

**Reply by 1111HGH** [reply]
or maybe you can combine Numpad6 and Right see them as the same key?(Numpad 1~9 as well) 
I mean if you release Numpad6, at the time, you make a judge: if right is pressed(I mean the other kind of Numpad6), then cancel display the'Numpad6' 

If you release 'Right', at the time, make a judge: if Numpad6 is pressed then cancel display the'Numpad6' 

And this way may cause more bugs, but that's suitable for some people, few people will press Right and Numpad6 at one time hahaha.(Maybe fewer people will do the things what I do hhh)

Or you can add a button into the window for the new mode.

;OK, that just an advise. It's my honor to meet such a polite writer when I first used the github. Thank you, take care of yourself!
